astropy/time/tests/test_custom_formats.py

```diff
@@ -0,0 +1,26 @@
+import pytest
+
+from astropy.time import Time, TimeFormat
+
+
+def test_custom_time_format_exception():
+    class SpecificException(ValueError):
+        pass
+
+    # Note there is no way to remove this after the test!
```

Comment:
What do you mean? Can you `del Custom` in [module teardown](https://docs.pytest.org/en/latest/xunit_setup.html#module-level-setup-teardown) ?

Reply:
Agreed with @pllim, though I would make this a `class TestCustomTimeFormat`, so that you can have easy `setup` and `teardown` methods.  In `setup(self)', you can then define the format, and in `teardown` remove it, using
```
def setup(self):
    class Custom(...):
        ...

    self.Custom  = Custom

def self.teardown(self):
    Time.FORMATS.pop(self.Custom)
```
This has the advantage that we can (later) expand this class to test a custom class more thoroughly...

Reply:
Sorry, my example was somewhat wrong - edited.

Reply:
It was the need to know about metaclass internals to deregister that I was referring to. And, more specifically, the need to make sure that different tests used different string names even though they were locally-defined classes. But de-registering them seems like a good trade-off.

Incidentally the register-when-class-created pattern seems to be the best available but it leads to some awkward situations where whether or not you imported a module affects the behaviour of code that uses no symbols from that module. In PINT my answer was to replace of "from astropy.time import Time" with "from pulsar_mjd import Time" so there's no reasonable way of constructing a Time object without first making available format="pulsar_mjd".

Reply:
I concluded two things:

- At least for these exception-raising tests, every test needs a custom time format that behaves differently. So by far the easiest way to do this is to construct a custom class in each test that needs one. This is kind of good anyway because custom formats shouldn't need a lot of boilerplate.
- Successfully installing and removing such a class is awkward and annoying, and the easiest place to put it was in a pytest fixture that returns a safe name and then deletes it after the test.